internal/metastore/table/table_catalog.go

```diff
@@ -2,88 +2,939 @@ package table
 }
 
-func (tc *Catalog) DropIndex(ctx context.Context, collectionInfo *model.Collection, dropIdxID typeutil.UniqueID, ts typeutil.Timestamp) error {
+func (tc *Catalog) AlterIndex(ctx context.Context, oldIndex *model.Index, newIndex *model.Index, alterType metastore.AlterType) error {
+	// no need updating table indexes
+	sqlStr := "update segment_indexes set partition_id=?, build_id=?, enable_index=?, index_file_paths=?, index_size=? where collection_id=? and segment_id=? and field_id=? and index_id=?"
+	tenantID := contextutil.TenantID(ctx)
+	for _, segIndex := range newIndex.SegmentIndexes {
+		indexFilePaths, err := json.Marshal(segIndex.IndexFilePaths)
+		if err != nil {
+			log.Error("marshal alias failed", zap.Error(err))
+			continue
+		}
+		indexFilePathsStr := string(indexFilePaths)
+		args := []interface{}{segIndex.PartitionID, segIndex.BuildID, segIndex.EnableIndex, indexFilePathsStr, segIndex.IndexSize, oldIndex.CollectionID, segIndex.SegmentID, oldIndex.FieldID, oldIndex.IndexID}
+		if tenantID != "" {
+			sqlStr = sqlStr + " and tenant_id=?"
+			args = append(args, tenantID)
+		}
+		_, err = tc.DB.Exec(sqlStr, args...)
+		if err != nil {
+			log.Error("update segment_indexes failed", zap.Error(err))
+			return err
+		}
+	}
+
 	return nil
 }
 
+func (tc *Catalog) DropIndex(ctx context.Context, collectionInfo *model.Collection, dropIdxID typeutil.UniqueID) error {
+	tenantID := contextutil.TenantID(ctx)
+	return WithTransaction(tc.DB, func(tx Transaction) error {
+		// sql 1
+		sqlStr1 := "update indexes set is_deleted=true where index_id=?"
+		args1 := []interface{}{dropIdxID}
+		if tenantID != "" {
+			sqlStr1 = sqlStr1 + " and tenant_id=?"
+			args1 = append(args1, tenantID)
+		}
+		rs, err := tx.Exec(sqlStr1, args1...)
+		if err != nil {
+			log.Error("update indexes by index ID failed", zap.Error(err), zap.Int64("indexID", dropIdxID))
+			return err
+		}
+		n, err := rs.RowsAffected()
+		if err != nil {
+			log.Error("get RowsAffected failed", zap.Error(err))
+			return err
+		}
+		log.Debug("table indexes RowsAffected", zap.Any("rows", n))
+
+		// sql 2
+		sqlStr2 := "update segment_indexes set is_deleted=true where index_id=?"
+		args2 := []interface{}{dropIdxID}
+		if tenantID != "" {
+			sqlStr2 = sqlStr2 + " and tenant_id=?"
+			args2 = append(args2, tenantID)
+		}
+		rs, err = tx.Exec(sqlStr2, args2...)
+		if err != nil {
+			log.Error("update segment_indexes by index ID failed", zap.Error(err), zap.Int64("indexID", dropIdxID))
+			return err
+		}
+		n, err = rs.RowsAffected()
+		if err != nil {
+			log.Error("get RowsAffected failed", zap.Error(err))
+			return err
+		}
+		log.Debug("table segment_indexes RowsAffected", zap.Any("rows", n))
+
+		return nil
+	})
+}
+
+func (tc *Catalog) listIndexesByCollectionID(ctx context.Context, collID typeutil.UniqueID) (map[int64]Index, error) {
+	var resultMap map[int64]Index
+
+	sqlStr := `select * from indexes where is_deleted=false`
+	args := []interface{}{}
+	if collID > 0 {
```

Comment:
why may the collection id be less than 0?

Reply:
Because it also used by list collection api.